File: apps/elixir/lib/getting_started_ml/chapter15/model_store.ex
defmodule GettingStartedMl.Chapter15.ModelStore do
  @moduledoc """
  学習済みモデルの置き場の約束。Java 版の `interface ModelStore`、
  Clojure 版の `defprotocol ModelStore` に当たる。

  Elixir で「約束」を表すのは **behaviour**（`@callback` を並べたモジュール）である。
  behaviour はモジュールに対する約束なので、置き場そのものは
  `{モジュール, 状態}` の組で表し、この組を受け取る関数で振り分ける。
  Plug が `{Plug のモジュール, 初期化した値}` を持ち回るのと同じ形である。

  behaviour が決めるのは **関数の名前と引数の数** だけで、
  「読み込めなければ ModelNotFoundError を投げる」という取り決めは書けない
  （テストの `StoreContract.check/2` で、本物と偽物の両方に確かめる）。
  """

  @typedoc "置き場。behaviour を実装したモジュールと、その状態の組。"
  @type t :: {module(), term()}

  @doc "映画の特徴量から興行収入を返す関数を読み込む。無ければ ModelNotFoundError を投げる。"
  @callback load_sales_model(state :: term()) :: (map() -> number())

  @doc "乗客の特徴量から生存するかどうかを返す関数を読み込む。無ければ ModelNotFoundError を投げる。"
  @callback load_survival_model(state :: term()) :: (map() -> boolean())

  @doc "置き場から、興行収入のモデルを読み込む。"
  def load_sales_model({module, state}), do: module.load_sales_model(state)

  @doc "置き場から、生存予測のモデルを読み込む。"
  def load_survival_model({module, state}), do: module.load_survival_model(state)
end

File: apps/php/src/Chapter15/ModelStore.php
<?php

declare(strict_types=1);

namespace GettingStartedMl\Chapter15;

/**
 * 学習済みモデルの置き場の約束。
 *
 * Elixir 版の behaviour は「モジュールへの約束」だったので、置き場を
 * `{モジュール, 状態}` の組で表す必要があった。**PHP の interface は値への約束** なので、
 * 置き場はただのオブジェクトでよく、状態（ディレクトリの名前）はそのオブジェクトが持つ。
 *
 * interface が決めるのは **メソッドの名前と引数と戻り値の型** だけで、
 * 「読み込めなければ ModelNotFoundException を投げる」という取り決めは書けない
 * （テストの `StoreContract::check()` で、本物と偽物の両方に確かめる）。
 */
interface ModelStore
{
    /** 興行収入のモデルを読み込む。無ければ ModelNotFoundException を投げる。 */
    public function loadSalesModel(): SalesModel;

    /** 生存予測のモデルを読み込む。無ければ ModelNotFoundException を投げる。 */
    public function loadSurvivalModel(): SurvivalModel;
}
